circuit_knitting/cutting/cutting_decomposition.py

```diff
@@ -260,3 +270,119 @@ def decompose_observables(
+              operations that can be performed by the search algorithm.
+            - seed: A seed for the pseudorandom number generator used by the optimizer
+        constraints: Options for specifying the constraints for circuit cutting
+
+            - qubits_per_QPU: The maximum number of qubits each subcircuit can contain
+              after cutting.
+
+    Returns:
+        A circuit containing :class:`.BaseQPDGate` instances. The subcircuits
+        resulting from cutting these gates will be runnable on the devices
+        specified in ``constraints``.
+
+        A metadata dictionary:
+            - cuts: A list of length-2 tuples describing each cut in the output circuit.
+              The tuples are formatted as ``(cut_type: str, cut_id: int)``. The
+              cut ID is the index of the cut gate or wire in the output circuit's
+              ``data`` field.
+            - sampling_overhead: The sampling overhead incurred from cutting the specified
+              gates and wires.
+
+    Raises:
+        ValueError: The input circuit contains a gate acting on more than 2 qubits.
+    """
+    circuit_cco = qc_to_cco_circuit(circuit)
+    interface = SimpleGateList(circuit_cco)
+
+    opt_settings = OptimizationSettings(
+        seed=optimization.seed,
+        max_gamma=optimization.max_gamma,
+        max_backjumps=optimization.max_backjumps,
+    )
+
+    # Hard-code the optimizer to an LO-only optimizer
+    optimizer = LOCutsOptimizer(interface, opt_settings, constraints)
+
+    # Find cut locations
+    opt_out = optimizer.optimize()
+
+    wire_cut_actions = []
+    gate_ids = []
+
+    opt_out = cast(DisjointSubcircuitsState, opt_out)
+    opt_out.actions = cast(list, opt_out.actions)
+    for action in opt_out.actions:
+        if action.action.get_name() == "CutTwoQubitGate":
+            gate_ids.append(action.gate_spec.instruction_id)
+        else:
+            # The cut-finding optimizer currently only supports 4 cutting
+            # actions: {CutTwoQubitGate + these 3 wire cut types}
+            assert action.action.get_name() in (
+                "CutLeftWire",
+                "CutRightWire",
+                "CutBothWires",
+            )
+            wire_cut_actions.append(action)
+
+    # First, replace all gates to cut with BaseQPDGate instances.
+    # This assumes each gate to cut is replaced 1-to-1 with a QPD gate.
+    # This may not hold in the future as we stop treating gate cuts individually.
+    circ_out = cut_gates(circuit, gate_ids)[0]
+
+    # Insert all the wire cuts
+    counter = 0
+    for action in sorted(wire_cut_actions, key=lambda a: a[1][0]):
+        inst_id = action.gate_spec.instruction_id
+        # args[0][0] will be either 1 (control) or 2 (target)
+        qubit_id = action.args[0][0] - 1
+        circ_out.data.insert(
+            inst_id + counter,
+            CircuitInstruction(
+                CutWire(), [circuit.data[inst_id].qubits[qubit_id]], []
```

Comment:
Let's make a test case that would fail if this still said `inst_id + counter` rather than `inst_id` (fixed in 25a9d06ddc31d32ddc119ac36cdff74e8a48eb01).

Reply:
'tis done, [f91fac0](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/pull/471/commits/f91fac09f6bceef223943d53d68e3d3f60e24b70).